lib/internal/streams/operators.js

```diff
@@ -0,0 +1,125 @@
+'use strict';
+
+const { AbortController } = require('internal/abort_controller');
+const { AbortError } = require('internal/errors');
+const Readable = require('internal/streams/readable');
+const eos = require('internal/streams/end-of-stream');
+
+module.exports.map = function map(stream, fn, options) {
+  // TODO: Argument validation
+
+  let concurrency = 1;
+  if (Number.isFinite(options)) {
+    concurrency = options;
+  } else if (options && Number.isFinite(options.concurrency)) {
+    concurrency = options.concurrency;
+  }
+
+  let highWaterMark = 1;
+  if (options && Number.isFinite(options.highWaterMark)) {
+    highWaterMark = options.highWaterMark;
+  }
+  highWaterMark = Math.max(0, highWaterMark - concurrency)
+
+  let objectMode = stream.readableObjectMode ?? stream.objectMode ?? true;
+  if (options && typeof options.objectMode === 'boolean') {
+    objectMode = options.objectMode;
+  }
+
+  const ac = new AbortController();
+  const signal = ac.signal;
+  const queue = [];
+
+  let reading = false;
+
+  const ret = new Readable({
+    objectMode,
+    highWaterMark,
+    read () {
+      if (!reading) {
+        read();
+      }
+    },
+    destroy (err, callback) {
+      if (!err && !this.readableEnded) {
+        err = new AbortError();
+      }
+      ac.abort();
+      callback(err);
+    }
+  });
+
+  async function read () {
+    if (ret.readableLength && ret.readableLength >= ret.readableHighWaterMark) {
+      return;
+    }
+
+    try {
+      reading = true;
+      while (queue.length && !ret.destroyed) {
+        let entry = queue.shift();
+
+        if (typeof entry.then === 'function') {
+          entry = await entry;
+        }
+
+        const [err, val] = entry;
+        if (err) {
+          ret.destroy(err);
+        } else if (!ret.push(val)) {
+          break;
+        } else {
+          pump();
+        }
+      }
+      reading = false;
+    } catch (err) {
+      ret.destroy(err);
+    }
+  }
+
+  async function wrap (val) {
+    try {
+      return [null, await val];
+    } catch (err) {
+      return [err, null];
+    }
+  }
+
+  function enqueue(val) {
+    queue.push(val);
+    if (!reading) {
+      read();
+    }
+  }
+
+  function pump () {
+    while (queue.length < concurrency) {
+      let val = stream.read();
+      if (val === null) {
+        return;
+      }
+
+      try {
+        val = fn(val, { signal });
+        if (val && typeof val.then === 'function') {
+          enqueue(wrap(val));
```

Comment:
I don't understand `wrap` is it just for unhandled rejections? (You are pushing a promise anyway right? Wouldn't it be simpler to `try/catch` on the consuming side and `.catch(() => {})` if the goal is to just suppress the unhandled rejections?

Reply:
We also need to propagate the error...